### PR TITLE
Update local algorithm with starting threshold and deviation threshold

### PR DESCRIPTION
**Pull request by ckyiricky**
Update local algorithm with starting threshold and deviation threshold
Change score mechanism to 45, 40, 15
Upload aggregate results for algs with large scaled dataset

### CODE REVIEW

File: modeling/algorithm/local-slice-algorithm.go [hunk 1]
```diff
@@ -59,86 +83,128 @@ func (alg LocalSliceAlgorithm) CreateSliceGroups(region types.RegionInfo) (map[s
...
-			weightedEndpoints.Number = zone.Endpoints
+		if zone.Endpoints != 0 {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
 		}
-		localGroup.Composition[zoneName] = weightedEndpoints
 		sliceGroups[zoneName] = localGroup
-	}
 
-	err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, sliceGroups)
+		// if this zone has a deviation below threshold after giving one
```
Comment (robscott) [inline modeling/algorithm/local-slice-algorithm.go]:
```suggestion
		// if this zone would still have a deviation below threshold after giving one
```

File: modeling/algorithm/local-slice-algorithm.go [hunk 2]
```diff
@@ -59,86 +83,128 @@ func (alg LocalSliceAlgorithm) CreateSliceGroups(region types.RegionInfo) (map[s
...
+// detect whether a zone is valid to contribute endpoints to other zones
+func (alg LocalSliceAlgorithm) validContributor(zoneName string, region types.RegionInfo, sliceGroups map[string]types.EndpointSliceGroup) bool {
+	// if the sliceGroup has no local composition, it is not a valid contributor
+	if len(sliceGroups[zoneName].Composition) == 0 || sliceGroups[zoneName].NumberOfEndpoints() <= 1 {
+		return false
+	}
+	return !alg.deviationAboveThreshold(zoneName, region, sliceGroups, -1)
+}
+
+// check if endpoints in a zone have invalid deviation
```
Comment (robscott) [inline modeling/algorithm/local-slice-algorithm.go:205]: Add quick comment about what delta means.

File: modeling/algorithm/local-slice-algorithm_test.go [hunk 3]
```diff
@@ -24,6 +24,159 @@ import (
 
 func TestLocalAlgorithm(t *testing.T) {
 	testCases := []algTestCase{
+		{
+			name: "corner case 1",
```
Comment (robscott) [inline modeling/algorithm/local-slice-algorithm_test.go]: add a bit more info for why this is/was a corner case

File: modeling/algorithm/utils.go [hunk 4]
```diff
@@ -136,3 +136,28 @@ func sortZoneByNames(zones map[string]types.Zone) []string {
 	sort.Strings(names)
 	return names
 }
+
+// assignEndpoints helps distribute endpoints from rich zones to poor zones in
+// local based algorithms
+func assignEndpoints(receiveZone *endpointDeviation, endpointsAvailable *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) {
```
Comment (robscott) [inline modeling/algorithm/utils.go:142]: Maybe add a bit more comments here.

**Review by robscott — COMMENTED** [review]
Thanks for your work on this!

File: modeling/algorithm/local-slice-algorithm.go (hunk 1 above)
Comment (ckyiricky) [reply modeling/algorithm/local-slice-algorithm.go]: udpated

File: modeling/algorithm/local-slice-algorithm.go (hunk 2 above)
Comment (ckyiricky) [reply modeling/algorithm/local-slice-algorithm.go:205]: added

File: modeling/algorithm/local-slice-algorithm_test.go (hunk 3 above)
Comment (ckyiricky) [reply modeling/algorithm/local-slice-algorithm_test.go]: changed case name

File: modeling/algorithm/utils.go (hunk 4 above)
Comment (ckyiricky) [reply modeling/algorithm/utils.go:142]: added

File: modeling/algorithm/local-slice-algorithm.go [hunk 5]
```diff
@@ -59,86 +83,133 @@ func (alg LocalSliceAlgorithm) CreateSliceGroups(region types.RegionInfo) (map[s
...
+// zero delta : if current state is above threshold
+// negative delta: after giving out abs(delta) endpoints, if it is still above
+// threshold
+// positive delta: after receiving delta endpoints, if it is sill above
```
Comment (robscott) [inline modeling/algorithm/local-slice-algorithm.go]: Tiny nit:
```suggestion
// positive delta: after receiving delta endpoints, if it is still above
```

Comment (ckyiricky) [reply modeling/algorithm/local-slice-algorithm.go]: Thanks! Changed.

**Review by robscott — APPROVED** [review]
Thanks!

**Pull request merged**